zqd/ingest/logreader.go

```diff
@@ -0,0 +1,124 @@
+package ingest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"sync/atomic"
+
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zio/ndjsonio"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zqe"
+)
+
+type MultipartLogReader struct {
+	mr       *multipart.Reader
+	opts     zio.ReaderOpts
+	stopErr  bool
+	warnings []string
+	zreader  *zbuf.File
+	zctx     *resolver.Context
+	nread    int64
+	done     bool
+}
+
+func NewMultipartLogReader(mr *multipart.Reader, zctx *resolver.Context) *MultipartLogReader {
+	return &MultipartLogReader{
+		mr:   mr,
+		opts: zio.ReaderOpts{Zng: zngio.ReaderOpts{Validate: true}},
+		zctx: zctx,
+	}
+}
+
+func (m *MultipartLogReader) SetStopOnError() {
+	m.stopErr = true
+}
+
+func (m *MultipartLogReader) Read() (*zng.Record, error) {
+read:
+	if m.done {
+		return nil, nil
+	}
+	if m.zreader == nil {
+		zr, err := m.next()
+		if err == io.EOF {
+			m.done = true
+			return nil, nil
+		}
```

Comment:
```suggestion
	if m.zreader == nil {
		zr, err := m.next()
		if err == io.EOF {
			return nil, nil
		}
```
(and if you apply this, also remove `done` from `MultipartLogReader`.

Reply:
You know I had this but for some reason I was getting an extra Read call after EOS. I was too lazy to debug it, but I'll take a second 👀

Reply:
We need keep this. `ReadBatch` for scanner requires receiving two EOS Reads from the underlying reader. One two prematurely close the `Batch` and another to return `EOS` on the next call. For some reason the golang multipart reader returns a different, unwrapped `EOF` after the first `EOF` is returned. (see https://golang.org/src/mime/multipart/multipart.go?s=8945:8987#L339). Frustrating. I will add a comment for clarity.

Reply:
Put a fix for this here: https://github.com/golang/go/pull/41560